notebooks/Numba_053_profiling_the_compiler.ipynb

```diff
@@ -0,0 +1,1045 @@
+       "          clearInterval(timer);\n",
+       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
+       "        }\n",
+       "      }\n",
+       "    }, 10, root)\n",
+       "  }\n",
+       "})(window);"
+      ]
+     },
+     "metadata": {
+      "application/vnd.bokehjs_exec.v0+json": {
+       "id": "1003"
+      }
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "p = figure(y_range=source['events'],\n",
+    "           plot_width=800, plot_height=600, toolbar_location=None,\n",
+    "           title=\"Compilation events\")\n",
+    "p.hbar(y=\"events\", left='start_times', right='end_times', height=.8, source=source)\n",
+    "p.xaxis.axis_label = \"Seconds since first event\"\n",
+    "\n",
+    "show(p)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The plot above shows all the compilation events triggered for `foo()`. We can see that it triggers 19 functions to be compiled."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The event API is used internally to record time spent in the compiler and LLVM. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Timers: {'llvm_lock': 1.1270671279999966, 'compiler_lock': 2.46969849}\n",
+      "\n",
+      "Total time spent in LLVM: 1.13s\n",
+      "Total time spent in the compiler (includes LLVM time): 2.47s\n",
+      "Percent of time spent in LLVM: 45.6%\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Print out the timers for compiler_lock\n",
+    "md = foo.get_metadata(foo.signatures[0])\n",
+    "timers = md['timers']\n",
+    "print(f\"Timers: {timers}\")\n",
+    "print()\n",
+    "print(f\"Total time spent in LLVM: {timers['llvm_lock']:.2f}s\")\n",
+    "print(f\"Total time spent in the compiler (includes LLVM time): {timers['compiler_lock']:.2f}s\")\n",
+    "\n",
+    "print(f\"Percent of time spent in LLVM: {timers['llvm_lock'] / timers['compiler_lock'] * 100:.1f}%\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As we can see, about half of the time is spent in LLVM. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## LLVM Pass Timings"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To understand why some functions take a long time in LLVM, we can look at the [LLVM pass timings](https://numba.readthedocs.io/en/0.53.0/developer/llvm_timings.html) to see what passes are taking a long time.\n",
+    "\n",
+    "This feature is turned off by default because of the overhead for LLVM to record the pass timings. To turn this on, set `numba.core.config.LLVM_PASS_TIMINGS` to truthy as done at the start of this notebook.\n",
```

Comment:
```suggestion
    "This feature is turned off by default because of the overhead incurred in LLVM when recording the pass timings. To turn this on, set `numba.core.config.LLVM_PASS_TIMINGS` to truthy as done at the start of this notebook.\n",
```